src/transformers/generation/utils.py

```diff
@@ -1741,6 +1743,27 @@ def generate(
                 model_kwargs["past_key_values"] = self._get_cache(
                     generation_config.cache_implementation, batch_size, generation_config.max_length
                 )
+            elif generation_config.cache_implementation == "mamba":
+                from ..models.mamba.modeling_mamba import MambaCache, MambaConfig
+
+                if not isinstance(self.config, MambaConfig):
+                    raise ValueError(
+                        "You can only specify `cache_implementation` to `mamba` if you are using mamba model"
+                    )
+
+                if hasattr(self, "_cache"):
+                    assert isinstance(self._cache, MambaCache), "Only `MambaCache` can be used on mamba model"
+                    need_new_cache = self._cache.conv_states.shape[1] != batch_size
+                else:
+                    need_new_cache = True
+
+                if need_new_cache:
+                    self._cache = MambaCache(
+                        config=self.config, batch_size=batch_size, dtype=self.dtype, device=self.device
+                    )
+                else:
+                    self._cache.reset()
+                model_kwargs["cache_params"] = self._cache
```

Comment:
isolate mamba cache setting logic seems to be the best way of keeping `_get_cache` clean, they need different number of paramters to initialize and different key names in `model_kwargs`